### PR TITLE
Add fallback strategy to CircuitBreakerClient

### LINKED ISSUE #4800: Add fallback strategy to `CircuitBreakerClient`
If a circuit for `CircuitBreakerClient` becomes open, requests are rejected with:
- `FailFastException`
https://github.com/line/armeria/blob/ab7488bafc2912445f026fb6d02420f134937073/core/src/main/java/com/linecorp/armeria/client/circuitbreaker/DefaultCircuitBreakerClientHandler.java#L47-L49
- `UnprocessedRequestException(CallNotPermittedException)`
https://github.com/line/armeria/blob/ab7488bafc2912445f026fb6d02420f134937073/resilience4j2/src/main/java/com/linecorp/armeria/resilience4j/circuitbreaker/client/Resilience4JCircuitBreakerClientHandler.java#L114-L118

If users want to add a fallback logic when the request is failed by `CircuitBreakerClient`, it is possible to add a custom decorator that wraps `CircuitBreakerClient` to recover.
```java
WebClient
  .builder(...)
  .decorator(CircuitBreakerClient.newDecorator(...))
  .decorator((delegate, ctx, req) -> {
   try {
        delegate.execute(ctx, req);
    // If resilent4j's `CircuitBreaker` is used.
    } catch(UnprocessedRequestException ex) {
        if (ex.getCause() instanceof CallNotPermittedException) {
            // Recover the request.
        }
    }
  });
```

That said, it would be more convenient for users to set a fallback logic via a builder method `CircuitBreakerClient`.
```java
CircuitBreakerClient
  .builder(...)
  .recover((ctx, req) -> {
     /* fallback */ 
     return HttpResponse.of(...);
  });
```

We may add a new API to `CircuitBreakerClientHandler` to check which exception is a circuit breaker exception.
```java
interface CircuitBreakerClientHandler {
   default boolean isCircuitBreakerException(Exception ex) {
       return ex instanceof FailFastException;
   }
}

class Resilience4JCircuitBreakerClientHandler implements CircuitBreakerClientHandler {
    @Override
    public boolean isCircuitBreakerException(Exception ex) {
        return ex instanceof UnprocessedRequestException && 
               ex.getCause() instanceof CallNotPermittedException;
    }
}
```

### PR DESCRIPTION
**Pull request by my4-dev**
Motivation:

Described in #4800 

Modifications:

- Add `recover` method to `CircuitBreakerClientBuilder` in order to set a fallback logic via a builder method
- Add a new API to `CircuitBreakerClientHandler` to check which exception is a circuit breaker exception

Result:

- Closes #4800

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java [hunk 1]
```diff
@@ -286,24 +288,52 @@ public static CircuitBreakerClientBuilder builder(
...
+
+    @Override
+    public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Exception {
+        try {
+            return super.execute(ctx, req);
+        } catch (Exception ex) {
+            if (handler().isCircuitBreakerException(ex) && fallback.isPresent()) {
+                return fallback.get().apply(ctx, req);
+            }
+            throw ex;
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java:335]: Should we move this logic to the super class?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientBuilder.java [hunk 2]
```diff
@@ -34,6 +39,8 @@
 
     private final boolean needsContentInRule;
     private final int maxContentLength;
+    private Optional<BiFunction<ClientRequestContext, HttpRequest, HttpResponse>> fallback =
+            Optional.empty();
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientBuilder.java:43]: Ditto, we prefer using `@Nullable`.

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientHandler.java [hunk 3]
```diff
@@ -98,4 +98,13 @@ static CircuitBreakerClientHandler of(CircuitBreakerMapping mapping) {
      */
     @Nullable
     CircuitBreakerCallback tryRequest(ClientRequestContext ctx, Request req);
+
+    /**
+     * Determines if the given {@link Exception} is related to a circuit breaker.
+     * This method is invoked by the {@link CircuitBreakerClient} to determine if a
+     * fallback logic should be executed.
+     */
+    default boolean isCircuitBreakerException(Exception ex) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientHandler.java:107]:
```suggestion
    @UnstableApi
    default boolean isCircuitBreakerException(Exception ex) {
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientBuilder.java [hunk 4]
```diff
@@ -73,6 +80,17 @@ public Function<? super HttpClient, CircuitBreakerClient> newDecorator() {
         return this::build;
     }
 
+    /**
+     * Sets the {@link BiFunction}. This is invoked when adding the fallback strategy.
+     *
+     * @return {@code this} to support method chaining.
+     */
+    public CircuitBreakerClientBuilder recover(
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientBuilder.java:88]: Should we also add this API to `CircuitBreakerRpcClientBuilder`?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientBuilder.java [hunk 5]
```diff
@@ -73,6 +80,17 @@ public Function<? super HttpClient, CircuitBreakerClient> newDecorator() {
...
+            BiFunction<ClientRequestContext, HttpRequest, HttpResponse> fallback) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientBuilder.java:89]:
```suggestion
            BiFunction<? super ClientRequestContext, ? super HttpRequest, ? extends HttpResponse> fallback) {
```

**Comment by my4-dev** [comment]
@ikhoon : Thank you for your advice! I've added the recover method to CircuitBreakerRpcClientBuilder and applied your advice.

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClient.java [hunk 6]
```diff
@@ -50,30 +51,37 @@
...
     private final CircuitBreakerClientHandler handler;
 
     /**
      * Creates a new instance that decorates the specified {@link Client}.
      */
     AbstractCircuitBreakerClient(Client<I, O> delegate, CircuitBreakerClientHandler handler,
-                                 CircuitBreakerRule rule) {
-        this(delegate, handler, requireNonNull(rule, "rule"), null);
+                                 CircuitBreakerRule rule,
+                                 BiFunction<? super ClientRequestContext, ? super I, ? extends O> fallback) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClient.java:63]:
```suggestion
                                 @Nullable BiFunction<? super ClientRequestContext, ? super I, ? extends O> fallback) {
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClient.java [hunk 7]
```diff
@@ -50,30 +51,37 @@
...
     /**
      * Creates a new instance that decorates the specified {@link Client}.
      */
-    AbstractCircuitBreakerClient(Client<I, O> delegate, CircuitBreakerClientHandler handler,
-                                 CircuitBreakerRuleWithContent<O> ruleWithContent) {
-        this(delegate, handler, null, requireNonNull(ruleWithContent, "ruleWithContent"));
+    AbstractCircuitBreakerClient(
+            Client<I, O> delegate, CircuitBreakerClientHandler handler,
+            CircuitBreakerRuleWithContent<O> ruleWithContent,
+            BiFunction<? super ClientRequestContext, ? super I, ? extends O> fallback) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClient.java:73]:
```suggestion
            @Nullable BiFunction<? super ClientRequestContext, ? super I, ? extends O> fallback) {
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClient.java [hunk 8]
```diff
@@ -115,13 +124,24 @@ final CircuitBreakerRule fromRuleWithContent() {
...
-            return unwrap().execute(ctx, req);
+        try {
+            final CircuitBreakerCallback callback = handler.tryRequest(ctx, req);
+            if (callback == null) {
+                return unwrap().execute(ctx, req);
+            }
+            return doExecute(ctx, req, callback);
+        } catch (Exception ex) {
+            if (handler().isCircuitBreakerException(ex) && fallback != null) {
+                return fallback.apply(ctx, req);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClient.java:141]: It should be better to early check nullness. 
```suggestion
                final O res = fallback.apply(ctx, req);
                return requireNonNull(res, "fallback.apply() returned null");
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClientBuilder.java [hunk 9]
```diff
@@ -103,6 +110,21 @@ final CircuitBreakerClientHandler handler() {
         return handler;
     }
 
+    final BiFunction<? super ClientRequestContext, ? super I, ? extends O> fallback() {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClientBuilder.java:113]:
```suggestion
    @Nullable
    final BiFunction<? super ClientRequestContext, ? super I, ? extends O> fallback() {
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClientBuilder.java [hunk 10]
```diff
@@ -103,6 +110,21 @@ final CircuitBreakerClientHandler handler() {
...
+        return fallback;
+    }
+
+    /**
+     * Sets the {@link BiFunction}. This is invoked when adding the fallback strategy.
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClientBuilder.java:118]: Could we add an example code for the API usage?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java [hunk 11]
```diff
@@ -281,9 +281,19 @@ public static CircuitBreakerClientBuilder builder(
...
+        needsContentInRule = false;
+        maxContentLength = 0;
+    }
+
+    /**
+     * Creates a new instance that decorates the specified {@link HttpClient}.
+     */
+    CircuitBreakerClient(
+            HttpClient delegate, CircuitBreakerClientHandler handler, CircuitBreakerRule rule,
+            BiFunction<? super ClientRequestContext, ? super HttpRequest, ? extends HttpResponse> fallback) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java:295]:
```suggestion
            @Nullable BiFunction<? super ClientRequestContext, ? super HttpRequest, ? extends HttpResponse> fallback) {
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java [hunk 12]
```diff
@@ -293,15 +303,17 @@ public static CircuitBreakerClientBuilder builder(
...
     /**
      * Creates a new instance that decorates the specified {@link HttpClient}.
      */
-    CircuitBreakerClient(HttpClient delegate, CircuitBreakerClientHandler handler,
-                         CircuitBreakerRuleWithContent<HttpResponse> ruleWithContent, int maxContentLength) {
-        super(delegate, handler, ruleWithContent);
+    CircuitBreakerClient(
+            HttpClient delegate, CircuitBreakerClientHandler handler,
+            CircuitBreakerRuleWithContent<HttpResponse> ruleWithContent, int maxContentLength,
+            BiFunction<? super ClientRequestContext, ? super HttpRequest, ? extends HttpResponse> fallback) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java:315]:
```suggestion
            @Nullable BiFunction<? super ClientRequestContext, ? super HttpRequest, ? extends HttpResponse> fallback) {
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRpcClient.java [hunk 13]
```diff
@@ -144,7 +144,17 @@ public static CircuitBreakerRpcClientBuilder builder(
...
+        this(delegate, handler, requireNonNull(ruleWithContent, "ruleWithContent"), null);
+    }
+
+    /**
+     * Creates a new instance that decorates the specified {@link RpcClient}.
+     */
+    CircuitBreakerRpcClient(
+            RpcClient delegate, CircuitBreakerClientHandler handler,
+            CircuitBreakerRuleWithContent<RpcResponse> ruleWithContent,
+            BiFunction<? super ClientRequestContext, ? super RpcRequest, ? extends RpcResponse> fallback) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRpcClient.java:156]:
```suggestion
            @Nullable BiFunction<? super ClientRequestContext, ? super RpcRequest, ? extends RpcResponse> fallback) {
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRpcClientBuilder.java [hunk 14]
```diff
@@ -58,4 +60,11 @@ public CircuitBreakerRpcClientBuilder mapping(CircuitBreakerMapping mapping) {
     public CircuitBreakerRpcClientBuilder handler(CircuitBreakerClientHandler handler) {
         return (CircuitBreakerRpcClientBuilder) super.handler(handler);
     }
+
+    @Override
+    public CircuitBreakerRpcClientBuilder recover(
+            BiFunction<? super ClientRequestContext, ? super RpcRequest, ? extends RpcResponse> fallback) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRpcClientBuilder.java:66]:
```suggestion
            @Nullable BiFunction<? super ClientRequestContext, ? super RpcRequest, ? extends RpcResponse> fallback) {
```

**Review by ikhoon — COMMENTED** [review]
Mostly, looks good.

**Comment by my4-dev** [comment]
@ikhoon : Thanks, I've applied all of your advice.

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClientBuilder.java [hunk 15]
```diff
@@ -103,6 +110,31 @@ final CircuitBreakerClientHandler handler() {
...
+     *   .builder(...)
+     *   .recover((ctx, req) -> {
+     *       // fallback logic
+     *       return HttpResponse.of(...);
+     *   });
+     * }</pre>
+     *
+     * @return {@code this} to support method chaining.
+     */
+    public AbstractCircuitBreakerClientBuilder<I, O> recover(
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClientBuilder.java:132]:
```suggestion
    @UnstableApi
    public AbstractCircuitBreakerClientBuilder<I, O> recover(
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientBuilder.java [hunk 16]
```diff
@@ -86,6 +92,14 @@ public CircuitBreakerClientBuilder handler(CircuitBreakerClientHandler handler)
         return (CircuitBreakerClientBuilder) super.handler(handler);
     }
 
+    @Override
+    public CircuitBreakerClientBuilder recover(@Nullable BiFunction<? super ClientRequestContext,
+            ? super HttpRequest, ? extends HttpResponse> fallback) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientBuilder.java:97]:
```suggestion
    public CircuitBreakerClientBuilder recover(BiFunction<? super ClientRequestContext,
            ? super HttpRequest, ? extends HttpResponse> fallback) {
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRpcClientBuilder.java [hunk 17]
```diff
@@ -58,4 +63,13 @@ public CircuitBreakerRpcClientBuilder mapping(CircuitBreakerMapping mapping) {
     public CircuitBreakerRpcClientBuilder handler(CircuitBreakerClientHandler handler) {
         return (CircuitBreakerRpcClientBuilder) super.handler(handler);
     }
+
+    @Override
+    public CircuitBreakerRpcClientBuilder recover(
+            @Nullable BiFunction<? super ClientRequestContext, ? super RpcRequest, ? extends RpcResponse>
+                    fallback) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRpcClientBuilder.java:70]:
```suggestion
    public CircuitBreakerRpcClientBuilder recover(
            BiFunction<? super ClientRequestContext, ? super RpcRequest, ? extends RpcResponse>
                    fallback) {
```

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClient.java [hunk 18]
```diff
@@ -115,13 +125,25 @@ final CircuitBreakerRule fromRuleWithContent() {
...
-        if (callback == null) {
-            return unwrap().execute(ctx, req);
+        try {
+            final CircuitBreakerCallback callback = handler.tryRequest(ctx, req);
+            if (callback == null) {
+                return unwrap().execute(ctx, req);
+            }
+            return doExecute(ctx, req, callback);
+        } catch (Exception ex) {
+            if (handler().isCircuitBreakerException(ex) && fallback != null) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClient.java:141]: null checking would be cheaper than the combination of method invocations and type checking. 
```suggestion
            if (fallback != null && handler().isCircuitBreakerException(ex)) {
```

**Review by ikhoon — APPROVED** [review]
Thanks, @my4-dev! Left minor comments about annotations.

**Comment by my4-dev** [comment]
Thanks for reviewing. I fixed it!

**Review by jrhee17 — APPROVED** [review]
Looks good 👍    Thanks @my4-dev 🙇 👍 🚀

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientBuilder.java [hunk 19]
```diff
@@ -86,6 +91,14 @@ public CircuitBreakerClientBuilder handler(CircuitBreakerClientHandler handler)
         return (CircuitBreakerClientBuilder) super.handler(handler);
     }
 
+    @Override
+    public CircuitBreakerClientBuilder recover(BiFunction<? super ClientRequestContext, ? super HttpRequest,
+            ? extends HttpResponse> fallback) {
+        requireNonNull(fallback, "fallback");
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientBuilder.java:97]: How about removing this null check and adding it in `AbstractCircuitBreakerClientBuilder.recover()` instead?

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRpcClientBuilder.java [hunk 20]
```diff
@@ -58,4 +62,12 @@ public CircuitBreakerRpcClientBuilder mapping(CircuitBreakerMapping mapping) {
     public CircuitBreakerRpcClientBuilder handler(CircuitBreakerClientHandler handler) {
         return (CircuitBreakerRpcClientBuilder) super.handler(handler);
     }
+
+    @Override
+    public CircuitBreakerRpcClientBuilder recover(BiFunction<? super ClientRequestContext, ? super RpcRequest,
+            ? extends RpcResponse> fallback) {
+        requireNonNull(fallback, "fallback");
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerRpcClientBuilder.java:69]: ditto

**Review by minwoox — COMMENTED** [review]
Looks great! 
Left some nits. 😄

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClientBuilder.java [hunk 21]
```diff
@@ -103,6 +110,32 @@ final CircuitBreakerClientHandler handler() {
...
+    @UnstableApi
+    public AbstractCircuitBreakerClientBuilder<I, O> recover(
+            BiFunction<? super ClientRequestContext, ? super I, ? extends O> fallback) {
+        this.fallback = fallback;
+        return this;
+    }
+
     @Override
     public String toString() {
         return MoreObjects.toStringHelper(this).omitNullValues()
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClientBuilder.java]: `toString()` method is removed.

**Review by minwoox — APPROVED** [review]
Thanks a lot, @my4-dev! 👍 🙇

**Comment by minwoox** [comment]
Thanks, again @my4-dev!

**Pull request merged**